src/data-access/repositories/mongo/mongo-part-changed-listener.ts

```diff
@@ -0,0 +1,85 @@
+import { BaseMongoRepository } from './base-mongo-repository'
+import { DataChangedListener } from '../interfaces/data-changed-listener'
+import { MongoDatabase } from './mongo-database'
+import { MongoEntityConverter, MongoPart, MongoSegment } from './mongo-entity-converter'
+import {
+  ChangeStream,
+  ChangeStreamDeleteDocument,
+  ChangeStreamDocument,
+  ChangeStreamInsertDocument,
+  ChangeStreamOptions,
+  ChangeStreamReplaceDocument
+} from 'mongodb'
+import { MongoChangeEvent } from './mongo-enums'
+import { Part } from '../../../model/entities/part'
+import { PartRepository } from '../interfaces/part-repository'
+
+const PART_COLLECTION_NAME: string = 'parts'
+
+export class MongoPartChangedListener extends BaseMongoRepository implements DataChangedListener<Part> {
+
+  private onCreatedCallback: (part: Part) => void
+  private onUpdatedCallback: (part: Part) => void
+  private onDeletedCallback: (partId: string) => void
+
+  constructor(
+    mongoDatabase: MongoDatabase,
+    mongoEntityConverter: MongoEntityConverter,
+    private readonly partRepository: PartRepository
+  ) {
+    super(mongoDatabase, mongoEntityConverter)
+    mongoDatabase.onConnect(PART_COLLECTION_NAME, () => this.listenForChanges())
+  }
+
+  private listenForChanges(): void {
+    const options: ChangeStreamOptions = { fullDocument: 'updateLookup' }
+    const changeStream: ChangeStream = this.getCollection().watch<MongoSegment, ChangeStreamDocument<MongoSegment>>([], options)
+    changeStream.on('change', (change: ChangeStreamDocument<MongoPart>) => void this.onChange(change))
+    console.debug('### Listening for Part collection changes...')
+  }
+
+  private async onChange(change: ChangeStreamDocument<MongoPart>): Promise<void> {
+    switch (change.operationType) {
+      case MongoChangeEvent.INSERT: {
```

Comment:
This method has multiple responsibilities, as it both figures out what event it is, and then handles it.
Each case should be extracted to their own method, to get rid of these extra responsibilities.

The new method would look something along the lines of:
```typescript
...
case MongoChangeEvent.INSERT: {
  return this.onInsertedEvent(change as ChangeStreamInsertDocument<MongoPart>)
}
...
```

With luck we might not need to await the new method, allowing for some parallel processing of events.
Not entirely sure if not awaiting, could become an issue during the callbacks.

Reply:
The responsibility of this method is to facilitate what happens when an event is received. It gets the data from the event and call the apropiated callback (sometimes it also calls a repository).

You can't avoid "facilitating" methods. What you can do is keep them on the same abstraction level.

You are not really gaining much extracting those methods and since `onChange` is current less than thirty lines, I would say it's overkill to extract them. If `onChange` grew larger, then I would also be in favour of extracting.